### PR TITLE
Fix image cropping for non-admins

### LINKED ISSUE #8912: Editors unable to crop poster images or logos
## Bug Description

When logged in a user with the role of Editor, the user is unable to crop poster images or publisher logos. Poster images and publisher logos can only be set if the selected image is already the correct size.

## Expected Behaviour

Editors should be able to crop poster images prior to setting.

## Steps to Reproduce

1. Log in as a user with the role of Editor
2. Open the poster image or publisher logo modals
3. Select an image that is not the correct size for a poster image or publisher logo
4. Select a section to crop
5. Error will display

## Screenshots

Video: https://jmp.sh/dIsUe9v

Screenshot of error:

image-crop

## Additional Context

- Plugin Version: v1.10.1
- WordPress Version: 5.8

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

Users with editor role were not able to crop poster images

## Summary

Fixes an issue with image cropping for users without the `customize` capability.

Directly using `CustomizeImageCropper` will set the `wp_customize` param which only works for admins, but in our case the cropping should also work for non-admins.

## Relevant Technical Choices

Adds a custom cropper similar to `CustomizeImageCropper`.

## User-facing changes

N/A

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Log in as an editor
2. Create a new story
3. Try to set a poster image or publisher logo which requires cropping
4. Verify that cropping works.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8912

### CODE REVIEW

File: packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    return wp.ajax.post('crop-image', {
+      nonce: attachment.get('nonces').edit,
+      id: attachment.get('id'),
+      context: control.id,
+      cropDetails: cropDetails,
+    });
+  },
+});
+
+window.wp.media.controller.WordPressImageCropper = WordPressImageCropper;
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js]: Why set this to the window? Why not just export as default like everywhere else?

File: packages/story-editor/src/components/mediaPicker/useMediaPicker.js [hunk 2]
```diff
@@ -196,7 +197,7 @@ export default function useMediaPicker({
       };
 
       // Create the media frame.
-      const fileFrame = global.wp.media({
+      const fileFrame = window.wp.media({
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/mediaPicker/useMediaPicker.js]: What does this change effect?

File: packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ * @see https://github.com/WordPress/wordpress-develop/blob/baa5ba32d8698f6a6a5928bcbd4fea35bd47a8c1/src/js/media/controllers/customize-image-cropper.js
+ */
+const WordPressImageCropper = window.wp?.media.controller.Cropper.extend({
+  /**
+   * Posts the crop details to the admin.
+   *
+   * Uses crop measurements when flexible in both directions.
+   * Constrains flexible side based on image ratio and size of the fixed side.
+   *
+   * @since 4.3.0
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js]:
```suggestion
```

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Some nitpics, but otherwise looking good.

File: packages/story-editor/src/components/mediaPicker/useMediaPicker.js (hunk 2 above)
Comment (swissspidy) [reply packages/story-editor/src/components/mediaPicker/useMediaPicker.js]: Nothing. They're exactly the same. But webpack replaces `global` with `window` when bundling.

And in webpack v5 it actually deprecates using `global`. You can learn more about that in #5792 

I was just fixing this while being at it.

File: packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js (hunk 1 above)
Comment (swissspidy) [reply packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js]: I was debugging the Karma tests. Will probably revert again.

Comment (spacedmonkey) [reply packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js]: > I was debugging the Karma tests. Will probably revert again.

Can you explain this a little more?

Comment (swissspidy) [reply packages/story-editor/src/components/mediaPicker/WordPressImageCropper.js]: Well the Karma tests are failing due to some globals being undefined and I was trying to debug why this happens and to make the Karma tests pass.

If you have a hunch what might be the issue, feel free to jump in and help fix the tests.

File: packages/story-editor/src/components/checklist/karma/checklist.karma.js [hunk 4]
```diff
@@ -402,10 +402,10 @@ describe('Checklist integration - Card visibility', () => {
 
     class Library {}
 
-    class CustomizeImageCropper {}
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/checklist/karma/checklist.karma.js:405]:
```
class Cropper {
    extend(){}
}

media.controller = {
      Cropper,
      Library,
};
```

I think this is issue. Remember we call `new window.wp.media.controller.WordPressImageCropper({` So cropper needs to be a class.

Comment (spacedmonkey) [reply packages/story-editor/src/components/checklist/karma/checklist.karma.js:405]: I pushed it, let's see what happens.

Comment (spacedmonkey) [reply packages/story-editor/src/components/checklist/karma/checklist.karma.js:405]: https://github.com/google/web-stories-wp/pull/8924/commits/ca022fd57e3a45a5ef9e4a718353a917bdc8f732

Comment (swissspidy) [reply packages/story-editor/src/components/checklist/karma/checklist.karma.js:405]: Can you reply to the CLA bot too?

Looks like there are still `Uncaught TypeError: Cannot read properties of undefined (reading 'media')` errors unfortunately.

**Comment by spacedmonkey** [comment]
@googlebot I consent.

**Comment by spacedmonkey** [comment]
There is still one test failing.

**Comment by spacedmonkey** [comment]
```
 Checklist integration - Card visibility
    hasUploadMediaAction=true
      ✗ should show cards that require the `hasUploadMediaAction` permission
	Uncaught TypeError: _WordPressImageCropper__WEBPACK_IMPORTED_MODULE_7__.default is not a constructor thrown

[0902/183232.973404:INFO:CONSOLE(271674)] "Uncaught TypeError: _WordPressImageCropper__WEBPACK_IMPORTED_MODULE_7__.default is not a constructor", source: http://localhost:9876/base/packages/story-editor/src/karma-tests.js (271674)
[0902/183233.006957:INFO:CONSOLE(271674)] "Uncaught TypeError: _WordPressImageCropper__WEBPACK_IMPORTED_MODULE_7__.default is not a constructor", source: http://localhost:9876/base/packages/story-editor/src/karma-tests.js (271674)
```

**Comment by swissspidy** [comment]
@spacedmonkey tests are all passing now

**Review by spacedmonkey — APPROVED** [review]

**Comment by spacedmonkey** [comment]
> @spacedmonkey tests are all passing now

Well done, this PR was a pain. This PR needs QA.

**Pull request merged**